src/LabelGroup/LabelGroup.tsx

```diff
@@ -1,17 +1,365 @@
+      onClick={collapseInlineExpandedChildren}
+      icon={() => <DashIcon size="small" />}
+      size="small"
+      aria-label="Show less"
+      variant="invisible"
+    />
+  ) : hiddenItemIds.length ? (
+    <Button ref={expandButtonRef} variant="invisible" size="small" onClick={showAllTokensInline}>
+      <VisuallyHidden>Show all</VisuallyHidden>
+      <span aria-hidden="true">+{hiddenItemIds.length}</span>
+    </Button>
+  ) : null
+
+const OverlayToggle: React.FC<
+  React.PropsWithChildren<{
+    closeOverflowOverlay: () => void
+    expandButtonRef: React.RefCallback<HTMLButtonElement>
+    hiddenItemIds: string[]
+    isOverflowShown: boolean
+    openOverflowOverlay: () => void
+    overlayPaddingPx: number
+    overlayWidth?: number
+  }>
+> = ({
+  children,
+  closeOverflowOverlay,
+  expandButtonRef,
+  hiddenItemIds,
+  isOverflowShown,
+  openOverflowOverlay,
+  overlayPaddingPx,
+  overlayWidth,
+}) =>
+  hiddenItemIds.length ? (
+    <AnchoredOverlay
+      open={isOverflowShown}
+      onOpen={openOverflowOverlay}
+      onClose={closeOverflowOverlay}
+      width="auto"
+      height="auto"
+      align="start"
+      side="inside-right"
+      // expandButtonRef satisfies React.RefObject<HTMLButtonElement> because we manually set `.current` in the `useCallback` above
+      anchorRef={expandButtonRef as unknown as React.RefObject<HTMLButtonElement>}
+      anchorOffset={overlayPaddingPx * -1}
+      alignmentOffset={overlayPaddingPx * -1}
+      renderAnchor={props => (
+        <Button variant="invisible" size="small" {...props} ref={expandButtonRef}>
+          <VisuallyHidden>Show all</VisuallyHidden>
+          <span aria-hidden="true">+{hiddenItemIds.length}</span>
+        </Button>
+      )}
+    >
+      <Box alignItems="flex-start" display="flex" width={overlayWidth} padding={`${overlayPaddingPx}px`}>
+        <Box display="flex" flexWrap="wrap" sx={{gap: 1}}>
+          {children}
+        </Box>
+        <IconButton onClick={closeOverflowOverlay} icon={XIcon} aria-label="Close" variant="invisible" />
+      </Box>
+    </AnchoredOverlay>
+  ) : null
+
+// TODO: reduce re-renders
+const LabelGroup: React.FC<React.PropsWithChildren<LabelGroupProps>> = ({
+  children,
+  visibleChildCount,
+  overflowStyle,
+  sx: sxProp,
+}) => {
+  const containerRef = React.useRef<HTMLDivElement>(null)
+  const collapseButtonRef = React.useRef<HTMLButtonElement>(null)
+  const firstHiddenIndexRef = React.useRef<number | undefined>(undefined)
+  const [visibilityMap, setVisibilityMap] = React.useState<Record<string, boolean>>({})
+  const [isOverflowShown, setIsOverflowShown] = React.useState<boolean>(false)
+  const [buttonClientRect, setButtonClientRect] = React.useState<DOMRect>({
+    width: 0,
+    right: 0,
+    height: 0,
+    x: 0,
+    y: 0,
+    top: 0,
+    left: 0,
+    bottom: 0,
+    toJSON: () => undefined,
+  })
+
+  const {theme} = useTheme()
+
+  const overlayPaddingPx = parseInt(get('space.2')(theme), 10)
```

Comment:
FYI, this code may break when we switch to using CSS variables. Shouldn't be a problem now though.